### DISCUSSION TITLE
args with keys helper?

### DISCUSSION THREAD

**Discussion by reslear** [wagmi-dev/abitype#225] [Q&A] [answered]
is there a ready way to get args with keys at once? 

```ts
// "tokenId" | "reciever"
type Names = MintAbi['inputs'][number]['name']

// readonly [`0x${string}`, bigint]
export type MintAbiArgs = AbiParametersToPrimitiveTypes<MintAbi['inputs']>

```

like: 
```ts
export type MintAbiArgsObject = {
  tokenId: `0x${string}`
  reciever: bigint
}
```

ideally zod helper:

```ts
export const mintInputSchema = z.object({
  tokenId: Address,
  reciever: z.bigint(),
})
```

**Comment by tmm** [comment, marked as answer]
You can do something like this:

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzjAnmApnAggI2AAUBDKEkLGLKAGhTUxwOLIqpoGlgA7AE3qoM2PIVLlK1KABUIRKKGAxgANyzSm9AL5wAZlAgg4AchKFhWY0iQXRrCRygBnWfMXK1G7E4A8DM-bsUk5wWAAe1PwhUFgkfBA8ADbocAAULOJBNHAAZIhwPOwAXHBOMAo8AOZwWgCUANoAuoJwAZmSXLx8oRFYUXbtjtz8cAC8A2wdUMMCSAB8Y4gM9YWUcLytYpOOTisAriD4NI31xquWjY0lGdtSrgogSqrqTH6oqACiEeQAxjA+bVuNF2PAORygzXy5xK5xqcxaqEBDikMwYcyQWmstgASlgnHtEvBxjdkcD7u5nl48W84AB6WnvAB6AH5lgxUMh3u9oSYnH0+DQrFyhEwSqY+HwYk4nEL3loEUthQViiYYj9gGBgH0YLKuRYxXFJXiZeyagrOcKeaYQBA9jwdaaRdgxXteDAAEwAVgAbLqzQxGvMkEA)

```ts
import {
  type AbiParameter,
  type AbiParameterKind,
  type AbiParameterToPrimitiveType,
} from 'abitype'

type AbiParametersToPrimitiveTypes<
  abiParameters extends readonly (AbiParameter & { name: string })[],
  abiParameterKind extends AbiParameterKind = AbiParameterKind,
> = {
  [name in abiParameters[number]['name']]: AbiParameterToPrimitiveType<
    Extract<abiParameters[number], { name: name }>,
    abiParameterKind
  >
}

type Result = AbiParametersToPrimitiveTypes<
  //   ^? type Result = { sender: `0x${string}`; recipient: `0x${string}`; amount: bigint; }
  [
    { name: 'sender'; type: 'address' },
    { name: 'recipient'; type: 'address' },
    { name: 'amount'; type: 'uint256' },
  ]
>
```

Note that `AbiParameter`s are not required to have the `'name'` property so this will not work if `'name'` does not exist, e.g. `[{ type: 'address' }, { type: 'address' }, { type: 'uint256' }]`. To get this to work with Zod, you'll need to define you own type utility from converting from a primitive type to value.
